src/common/test/myProfileMocking.ts

```diff
@@ -155,3 +207,189 @@ export const getVerifiedData = (
+): 'primaryAddress' | 'primaryPhone' | 'primaryEmail' {
+  return dataType === 'emails'
+    ? 'primaryEmail'
+    : dataType === 'phones'
+    ? 'primaryPhone'
+    : 'primaryAddress';
+}
+
+function createEdge(
+  dataType: EditDataType,
+  nodeData: Partial<MultiItemProfileNode>
+): Mutable<InsertableEdge> {
+  const typename = getEdgeTypeName(dataType);
+  const node = {
+    ...createNewProfileNode(dataType),
+    ...nodeData,
+  };
+  return {
+    __typename: typename,
+    node,
+  } as InsertableEdge;
+}
+
+function setPrimaryNode(
+  profile: Mutable<ProfileData>,
+  dataType: EditDataType,
+  newPrimary: Mutable<MultiItemProfileNode> | null
+): void {
+  if (newPrimary) {
+    newPrimary.primary = true;
+  }
+  const propName = getPrimaryPropName(dataType);
+  Object.assign(profile, { [propName]: newPrimary });
+}
+
+export function cloneAndManipulateProfile(
+  profileData: ProfileData
+): ManipulationFunctions {
+  const currentProfile = _.cloneDeep(profileData) as Mutable<ProfileData>;
+  // added nodes must have empty id in formValues but id in the created GraphQL
+  // which requires same extra check ups and deletion:
+  const addedNodes: Set<string> = new Set();
+  return {
+    getProfile() {
+      return currentProfile;
+    },
+    add(dataType, data, index = -1) {
+      const edge = createEdge(dataType, data);
+      const target = getEdges(currentProfile, dataType);
+      if (!target) {
+        throw Error('Cannot add edge');
+      }
+      if (index === -1) {
+        target.push(edge);
+      } else {
+        target.splice(index, 0, edge);
+      }
+      addedNodes.add(`${dataType}_${data.id}`);
+      return this;
+    },
+    remove(dataType, prop) {
+      const nodeList = pickSources(currentProfile, dataType);
+      const index = findNode(
+        nodeList as MultiItemProfileNode[],
+        prop.id,
+        true
+      ) as number;
+      const target = getEdges(currentProfile, dataType);
+      if (index < 0 || !target) {
+        throw Error('Cannot find node to remove or edges');
+      }
+      target.splice(index, 1);
+      return this;
+    },
+    setPrimary(dataType, node) {
+      setPrimaryNode(currentProfile, dataType, node);
+      return this;
+    },
+    edit(dataType, node) {
+      const target = getEdges(currentProfile, dataType);
+      if (!target) {
+        throw Error('Cannot edit edges');
+      }
+      const targetEdge = target.find(edge => edge?.node?.id === node.id);
+      if (!targetEdge) {
+        throw Error('Cannot edit edge');
+      }
+      Object.assign(targetEdge.node, node);
+      return this;
+    },
+    setBasicData(basicData) {
+      Object.assign(currentProfile, basicData);
+      return this;
+    },
+    setAdditionalInformation(additionalInformation) {
+      Object.assign(currentProfile, additionalInformation);
+      return this;
+    },
+    getFormValues() {
+      const basicData = pickSources(
+        currentProfile,
+        basicDataType
```

Comment:
Does basicDataType refer to first name, last name and nickname? Would namesDataType describe it better?

Reply:
They are grouped in UI as "Perustiedot". That's why using "Basic data". It might get additional fields that are not names. Language was in it too, but moved to "Lisätiedot".